### ISSUE TITLE
Spark/Splink3 - Connected components does not converge

### ISSUE THREAD

**Issue by RobinL** [moj-analytical-services/splink#466]
There's some issue with caching in Spark causing problems

Link using duckdb:
```
Root rows 26
Completed iteration 1
Root rows 2
Completed iteration 2
Root rows 0
Completed iteration 3
```

<details>

```
# %load_ext autoreload
# %autoreload 2


from splink.duckdb.duckdb_linker import DuckDBLinker

# from try_settings import settings_dict
import pandas as pd


df_1 = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
# df_1.rename(columns={"unique_id": "uid"}, inplace=True)

df_2 = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")
# df_2.rename(columns={"unique_id": "uid"}, inplace=True)

from splink.comparison_library import exact_match, levenshtein

settings = {
    # "unique_id_column_name": "uid",
    "proportion_of_matches": 4 / 1000,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.first_name = r.first_name",
        "l.surname = r.surname",
        "l.email = r.email",
    ],
    "comparisons": [
        levenshtein("first_name", 2),
        exact_match("surname"),
        exact_match("dob"),
        exact_match("city", term_frequency_adjustments=True),
        exact_match("email"),
    ],
    "retain_matching_columns": True,
    "retain_intermediate_calculation_columns": True,
    "additional_columns_to_retain": ["group"],
    "max_iterations": 5,
    "em_convergence": 0.01,
}


from splink.duckdb.duckdb_linker import DuckDBLinker


linker = DuckDBLinker(df_1, settings, set_up_basic_logging=True)
c = linker._settings_obj.comparisons[0]


linker.estimate_u_using_random_sampling(target_rows=1e6)


df_predict = linker.predict()
df_predict.as_pandas_dataframe().sort_values(["unique_id_l", "unique_id_r"])

# linker.debug_mode = True
cc = linker.cluster_pairwise_predictions_at_threshold(
    df_predict, threshold_match_probability=0.5
)
cc_pd = cc.as_pandas_dataframe()
cc_pd
```
</details>

Same link using Spark:
```
INFO:splink.connected_components:Root rows 26
INFO:splink.connected_components:Completed iteration 1
INFO:splink.connected_components:Root rows 26
INFO:splink.connected_components:Completed iteration 2
INFO:splink.connected_components:Root rows 26
INFO:splink.connected_components:Completed iteration 3
INFO:splink.connected_components:Root rows 26                                   
INFO:splink.connected_components:Completed iteration 4
```

<details>

```
import logging


logging.basicConfig()
logger = logging.getLogger("splink")

logger.setLevel(logging.DEBUG)
# logger.setLevel(logging.INFO)


from splink.comparison_library import exact_match, levenshtein

settings = {
    "proportion_of_matches": 4 / 1000,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.first_name = r.first_name",
        "l.surname = r.surname",
        "l.email = r.email",
    ],
    "comparisons": [
        levenshtein("first_name", 2),
        exact_match("surname"),
        exact_match("dob"),
        exact_match("city", term_frequency_adjustments=True),
        exact_match("email"),
    ],
    "retain_matching_columns": True,
    "retain_intermediate_calculation_columns": True,
    "additional_columns_to_retain": ["group"],
    "max_iterations": 1,
    "em_convergence": 0.01,
}


from pyspark.context import SparkContext, SparkConf
from pyspark.sql import SparkSession
from splink.spark.spark_linker import SparkLinker


conf = SparkConf()
conf.set("spark.driver.memory", "12g")
conf.set("spark.sql.shuffle.partitions", "8")
conf.set("spark.default.parallelism", "8")

sc = SparkContext.getOrCreate(conf=conf)
spark = SparkSession(sc)
spark.sparkContext.setCheckpointDir("./tmp_checkpoints")
df = spark.read.csv("./tests/datasets/fake_1000_from_splink_demos.csv", header=True)


linker = SparkLinker(df, settings, break_lineage_method="checkpoint")

c = linker._settings_obj.comparisons[0]
c.as_dict


linker.estimate_u_using_random_sampling(target_rows=1e6)



df_predict = linker.predict()

df_predict.as_pandas_dataframe().sort_values(["unique_id_l", "unique_id_r"])

# import logging

# logging.basicConfig(
#     format="%(message)s",
# )
# splink_logger = logging.getLogger("splink")
# splink_logger.setLevel(9)

linker.cluster_pairwise_predictions_at_threshold(df_predict, 0.5)

```

</details>

**Comment by RobinL** [comment]
More testing:

<details>

```
%load_ext autoreload
%autoreload 2


import logging
import pandas as pd



from pyspark.context import SparkContext, SparkConf
from pyspark.sql import SparkSession
from splink.spark.spark_linker import SparkLinker, SparkDataframe

settings = {
    "unique_id_column_name": "uid",
    "proportion_of_matches": 0.01,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.col1 = r.col1",
    ],
    "comparisons": [
        {
            "comparison_levels": [
                {
                    "sql_condition": "col1_l IS NULL OR col1_r IS NULL",
                    "is_null_level": True,
                },
                {"sql_condition": "col1_l = col1_r"},
                {"sql_condition": "ELSE"},
            ]
        }
    ],
    "retain_matching_columns": True,
    "retain_intermediate_calculation_columns": True,
}

conf = SparkConf()
conf.set("spark.driver.memory", "12g")
conf.set("spark.sql.shuffle.partitions", "8")
conf.set("spark.default.parallelism", "8")

sc = SparkContext.getOrCreate(conf=conf)
spark = SparkSession(sc)
spark.sparkContext.setCheckpointDir("./tmp_checkpoints")


df = [
    {"uid": "0", "col1": "a"},
    {"uid": "1", "col1": "b"},
    {"uid": "2", "col1": "c"},
    {"uid": "3", "col1": "d"},
    {"uid": "4", "col1": "e"},
    {"uid": "5", "col1": "f"},
    {"uid": "6", "col1": "g"},
    {"uid": "7", "col1": "h"},
    {"uid": "8", "col1": "i"},
]
df = spark.createDataFrame(pd.DataFrame(df))


linker = SparkLinker(df, settings, break_lineage_method="checkpoint")

data = [
    {"uid_l": "0", "uid_r": "1", "match_probability": 1.0},
    {"uid_l": "0", "uid_r": "2", "match_probability": 1.0},
    {"uid_l": "2", "uid_r": "4", "match_probability": 1.0},
    {"uid_l": "4", "uid_r": "8", "match_probability": 1.0},
    {"uid_l": "7", "uid_r": "8", "match_probability": 1.0},
]

spark.createDataFrame(pd.DataFrame(data)).createOrReplaceTempView("df_predict")

linker._initialise_df_concat_with_tf()

splink_logger = logging.getLogger("splink")
splink_logger.setLevel(logging.DEBUG)
linker.debug_mode = False
df_predict = SparkDataframe("df_predict", "df_predict", linker)
df_cc = linker.cluster_pairwise_predictions_at_threshold(df_predict, 0.5)
df_cc.as_pandas_dataframe()

```
</details>

**Comment by RobinL** [comment]
Closed by #467
